File: scripts/unity-sampled-skeleton.cs
// Invoked in a temporary Unity 2022 project by unity-sampled-skeleton.mjs.
// Unity is the authority for legacy FBX import, controller binding, and skin
// deformation. This script records the imported mesh, bind pose, hierarchy,
// and exact local transforms at every authored frame; Node then writes those
// values directly to glTF without passing animation data through FBX again.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using UnityEditor;
using UnityEditor.SceneManagement;
using UnityEngine;

public static class RtsUnitySampledSkeleton
{
    [Serializable]
    private sealed class JobFile
    {
        public string prefabPath;
        public string controllerPath;
        public string outputPath;
        public List<ClipJob> clips = new List<ClipJob>();
    }

    [Serializable]
    private sealed class ClipJob
    {
        public string name;
        public string state;
        public int frames;
        public float frameRate;
    }

    [Serializable]
    private sealed class SampleFile
    {
        public string unityVersion;
        public List<NodeRecord> nodes = new List<NodeRecord>();
        public float[] restTransforms;
        public MeshRecord mesh;
        public List<ClipRecord> clips = new List<ClipRecord>();
    }

    [Serializable]
    private sealed class NodeRecord
    {
        public string name;
        public string path;
    }

    [Serializable]
    private sealed class ClipRecord
    {
        public string name;
        public int authoredFrames;
        public float frameRate;
        public float[] transforms;
    }

    [Serializable]
    private sealed class MeshRecord
    {
        public int rendererNode;
        public int[] boneNodes;
        public float[] vertices;
        public float[] normals;
        public float[] uv;
        public int[] triangles;
        public int[] boneIndices;
        public float[] boneWeights;
        public float[] bindPoses;
    }

    public static void Run()
    {
        try
        {
            var manifestPath = CommandLineValue("-rtsSampleManifest");
            if (string.IsNullOrEmpty(manifestPath) || !File.Exists(manifestPath))
                throw new FileNotFoundException("Unity sampled-skeleton manifest is absent", manifestPath);
            var job = JsonUtility.FromJson<JobFile>(File.ReadAllText(manifestPath));
            if (job == null || string.IsNullOrEmpty(job.prefabPath) ||
                string.IsNullOrEmpty(job.controllerPath) || string.IsNullOrEmpty(job.outputPath))
                throw new InvalidDataException("Unity sampled-skeleton manifest is malformed");
            if (job.clips == null || job.clips.Count == 0)
                throw new InvalidDataException("Unity sampled-skeleton manifest contains no clips");

            EditorSceneManager.NewScene(NewSceneSetup.EmptyScene, NewSceneMode.Single);
            var importer = AssetImporter.GetAtPath(job.prefabPath) as ModelImporter;
            if (importer == null) throw new FileNotFoundException("Geometry FBX importer is absent", job.prefabPath);
            if (!importer.isReadable)
            {
                importer.isReadable = true;
                importer.SaveAndReimport();
            }

            var prefab = AssetDatabase.LoadAssetAtPath<GameObject>(job.prefabPath);
            var controller = AssetDatabase.LoadAssetAtPath<RuntimeAnimatorController>(job.controllerPath);
            if (!prefab || !controller)
                throw new FileNotFoundException("Sampled-skeleton prefab or controller is absent");

            var root = UnityEngine.Object.Instantiate(prefab);
            root.name = "SampledSkeleton";
            root.transform.SetPositionAndRotation(Vector3.zero, Quaternion.identity);
            root.transform.localScale = Vector3.one;
            root.SetActive(true);
            // UnityEngine.Object has a fake-null state that C#'s ?? operator
            // does not observe. Some imported model prefabs expose that stale
            // Animator reference, so use Unity's overloaded truth test here.
            var animator = root.GetComponent<Animator>();
            if (!animator) animator = root.AddComponent<Animator>();
            animator.runtimeAnimatorController = controller;
            animator.applyRootMotion = false;
            animator.cullingMode = AnimatorCullingMode.AlwaysAnimate;
            animator.enabled = true;

            var transforms = root.GetComponentsInChildren<Transform>(true);
            var output = new SampleFile
            {
                unityVersion = Application.unityVersion,
                restTransforms = new float[transforms.Length * 10],
            };
            foreach (var transform in transforms)
            {
                output.nodes.Add(new NodeRecord
                {
                    name = transform.name,
                    path = AnimationUtility.CalculateTransformPath(transform, root.transform),
                });
            }
            for (var node = 0; node < transforms.Length; node++)
                WriteTransform(output.restTransforms, node * 10, transforms[node]);

            var renderers = root.GetComponentsInChildren<Renderer>(true)
                .Where(candidate => candidate is MeshRenderer || candidate is SkinnedMeshRenderer)
                .ToArray();
            if (renderers.Length != 1 || !(renderers[0] is SkinnedMeshRenderer renderer) || !renderer.sharedMesh)
                throw new InvalidDataException(
                    "Sampled-skeleton prefab must contain exactly one complete SkinnedMeshRenderer; found " +
                    string.Join(", ", renderers.Select(candidate =>
                        candidate.name + ":" + candidate.GetType().Name)));
            output.mesh = ReadMesh(renderer, transforms);

            foreach (var clip in job.clips)
            {
                if (string.IsNullOrEmpty(clip.name) || string.IsNullOrEmpty(clip.state) ||
                    clip.frames < 1 || !(clip.frameRate > 0))
                    throw new InvalidDataException("Sampled-skeleton clip timing is malformed");
                var record = new ClipRecord
                {
                    name = clip.name,
                    authoredFrames = clip.frames,
                    frameRate = clip.frameRate,
                    transforms = new float[(clip.frames + 1) * transforms.Length * 10],
                };
                animator.Rebind();
                animator.Play(clip.state, 0, 0f);
                animator.Update(0f);
                var state = animator.GetCurrentAnimatorStateInfo(0);
                if (state.shortNameHash != Animator.StringToHash(clip.state))
                    throw new InvalidDataException("Animator has no state named " + clip.state);
                for (var frame = 0; frame <= clip.frames; frame++)
                {
                    if (frame > 0) animator.Update(1f / clip.frameRate);
                    for (var node = 0; node < transforms.Length; node++)
                    {
                        var offset = (frame * transforms.Length + node) * 10;
                        WriteTransform(record.transforms, offset, transforms[node]);
                    }
                }
                output.clips.Add(record);
                Debug.LogFormat("SAMPLED {0}: {1} keys, {2} nodes", clip.name, clip.frames + 1, transforms.Length);
            }

            Directory.CreateDirectory(Path.GetDirectoryName(job.outputPath));
            File.WriteAllText(job.outputPath, JsonUtility.ToJson(output));
            Debug.Log("WROTE " + job.outputPath);
            EditorApplication.Exit(0);
        }
        catch (Exception exception)
        {
            Debug.LogException(exception);
            EditorApplication.Exit(1);
        }
    }

    private static MeshRecord ReadMesh(SkinnedMeshRenderer renderer, Transform[] transforms)
    {
        var source = renderer.sharedMesh;
        var vertices = source.vertices;
        var normals = source.normals;
        var uv = source.uv;
        var weights = source.boneWeights;
        var bindPoses = source.bindposes;
        if (normals.Length != vertices.Length || uv.Length != vertices.Length || weights.Length != vertices.Length)
            throw new InvalidDataException("Sampled-skeleton mesh attributes have inconsistent lengths");
        var record = new MeshRecord
        {
            rendererNode = Array.IndexOf(transforms, renderer.transform),
            boneNodes = renderer.bones.Select(bone => Array.IndexOf(transforms, bone)).ToArray(),
            vertices = new float[vertices.Length * 3],
            normals = new float[normals.Length * 3],
            uv = new float[uv.Length * 2],
            triangles = source.triangles,
            boneIndices = new int[weights.Length * 4],
            boneWeights = new float[weights.Length * 4],
            bindPoses = new float[bindPoses.Length * 16],
        };
        if (record.rendererNode < 0 || record.boneNodes.Any(node => node < 0) ||
            record.boneNodes.Length != bindPoses.Length)
            throw new InvalidDataException("Sampled-skeleton renderer has an invalid bone mapping");
        for (var index = 0; index < vertices.Length; index++)
        {
            record.vertices[index * 3 + 0] = vertices[index].x;
            record.vertices[index * 3 + 1] = vertices[index].y;
            record.vertices[index * 3 + 2] = vertices[index].z;
            record.normals[index * 3 + 0] = normals[index].x;
            record.normals[index * 3 + 1] = normals[index].y;
            record.normals[index * 3 + 2] = normals[index].z;
            record.uv[index * 2 + 0] = uv[index].x;
            record.uv[index * 2 + 1] = uv[index].y;
            record.boneIndices[index * 4 + 0] = weights[index].boneIndex0;
            record.boneIndices[index * 4 + 1] = weights[index].boneIndex1;
            record.boneIndices[index * 4 + 2] = weights[index].boneIndex2;
            record.boneIndices[index * 4 + 3] = weights[index].boneIndex3;
            record.boneWeights[index * 4 + 0] = weights[index].weight0;
            record.boneWeights[index * 4 + 1] = weights[index].weight1;
            record.boneWeights[index * 4 + 2] = weights[index].weight2;
            record.boneWeights[index * 4 + 3] = weights[index].weight3;
        }
        for (var matrix = 0; matrix < bindPoses.Length; matrix++)
            for (var column = 0; column < 4; column++)
                for (var row = 0; row < 4; row++)
                    record.bindPoses[matrix * 16 + column * 4 + row] = bindPoses[matrix][row, column];
        return record;
    }

    private static void WriteTransform(float[] values, int offset, Transform transform)
    {
        values[offset + 0] = transform.localPosition.x;
        values[offset + 1] = transform.localPosition.y;
        values[offset + 2] = transform.localPosition.z;
        values[offset + 3] = transform.localRotation.x;
        values[offset + 4] = transform.localRotation.y;
        values[offset + 5] = transform.localRotation.z;
        values[offset + 6] = transform.localRotation.w;
        values[offset + 7] = transform.localScale.x;
        values[offset + 8] = transform.localScale.y;
        values[offset + 9] = transform.localScale.z;
    }

    private static string CommandLineValue(string name)
    {
        var args = Environment.GetCommandLineArgs();
        for (var index = 0; index + 1 < args.Length; index++)
            if (args[index] == name) return args[index + 1];
        return null;
    }
}

File: scripts/unity-sampled-skeleton.mjs
/**
 * Build a standard skeletal GLB from Unity's imported mesh and sampled local
 * transforms. This bypasses FBX export entirely: Unity owns the legacy FBX and
 * Animator semantics, while Three only serializes already-resolved glTF data.
 */

import { spawn } from 'node:child_process';
import { access, copyFile, mkdir, mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { basename, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import * as THREE from 'three';
import { GLTFExporter } from 'three/examples/jsm/exporters/GLTFExporter.js';

const UNITY_VERSION = '2022.3.62f3';
const DEFAULT_UNITY = `C:\\Program Files\\Unity\\Hub\\Editor\\${UNITY_VERSION}\\Editor\\Unity.exe`;
const CLIP_NAMES = ['run', 'attack', 'die'];
const POSITION_EPSILON = 1e-5;
const SCALE_EPSILON = 1e-7;
const QUATERNION_EPSILON = 1e-12;

installBrowserShims();

/** Run Unity in an isolated copy of the source assets, then serialize its data. */
export async function buildUnitySampledSkeletonModel(sourceRoot, spec, timings, unityOverride) {
  const configuration = spec.unitySampledSkeleton;
  if (!configuration?.controller) {
    throw new Error(`${spec.unit}: Unity sampled-skeleton controller is absent`);
  }
  const unity = await firstAvailablePath([
    unityOverride ? resolve(unityOverride) : null,
    DEFAULT_UNITY,
  ]);
  if (!unity) {
    throw new Error(
      `${spec.unit} requires Unity ${UNITY_VERSION}; install that editor or pass --unity.`,
    );
  }

  const geometrySlot = configuration.geometry ?? spec.geometry ?? 'run';
  const geometryFile =
    spec.geometryFile ?? (geometrySlot in spec.files ? spec.files[geometrySlot] : null);
  if (!geometryFile) {
    throw new Error(`${spec.unit}: sampled-skeleton geometry ${geometrySlot} is absent`);
  }
  validateTimings(timings, spec.unit);

  const tempRoot = await mkdtemp(join(tmpdir(), 'rts-athena2-unity-skeleton-'));
  try {
    const assetRoot = join(tempRoot, 'Assets', 'Source');
    const editorRoot = join(tempRoot, 'Assets', 'Editor');
    const packagesRoot = join(tempRoot, 'Packages');
    const settingsRoot = join(tempRoot, 'ProjectSettings');
    const outputRoot = join(tempRoot, 'Output');
    await Promise.all(
      [assetRoot, editorRoot, packagesRoot, settingsRoot, outputRoot].map((path) =>
        mkdir(path, { recursive: true }),
      ),
    );

    const sourceNames = [
      ...new Set(
        [...Object.values(spec.files), spec.geometryFile, configuration.controller].filter(Boolean),
      ),
    ];
    const sourceFiles = sourceNames.map((name) => join(sourceRoot, name));
    await Promise.all(sourceFiles.flatMap((path) => [access(path), access(`${path}.meta`)]));
    await Promise.all(
      sourceFiles.flatMap((path) => [
        copyFile(path, join(assetRoot, basename(path))),
        copyFile(`${path}.meta`, join(assetRoot, `${basename(path)}.meta`)),
      ]),
    );

    const sourceScript = fileURLToPath(new URL('./unity-sampled-skeleton.cs', import.meta.url));
    const samplePath = join(outputRoot, 'unity-samples.json');
    const manifestPath = join(tempRoot, 'sample-job.json');
    const manifest = {
      prefabPath: unityAssetPath(geometryFile),
      controllerPath: unityAssetPath(configuration.controller),
      outputPath: samplePath,
      clips: CLIP_NAMES.map((name) => ({
        name,
        state: configuration.states?.[name] ?? name,
        frames: timings[name].frames,
        frameRate: timings[name].frameRate,
      })),
    };
    await Promise.all([
      copyFile(sourceScript, join(editorRoot, 'RtsUnitySampledSkeleton.cs')),
      writeFile(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`),
      writeFile(
        join(packagesRoot, 'manifest.json'),
        `${JSON.stringify({ dependencies: {} }, null, 2)}\n`,
      ),
      writeFile(
        join(settingsRoot, 'ProjectVersion.txt'),
        `m_EditorVersion: ${UNITY_VERSION}\n` + `m_EditorVersionWithRevision: ${UNITY_VERSION}\n`,
      ),
    ]);

    console.log(
      `Sampling ${spec.unit} mesh and animation transforms with Unity ${UNITY_VERSION}...`,
    );
    const unityLog = join(tempRoot, 'unity.log');
    try {
      await runProcess(unity, [
        '-batchmode',
        '-nographics',
        '-quit',
        '-projectPath',
        tempRoot,
        '-executeMethod',
        'RtsUnitySampledSkeleton.Run',
        '-rtsSampleManifest',
        manifestPath,
        '-logFile',
        unityLog,
      ]);
    } catch (error) {
      let tail = '';
      try {
        const log = await readFile(unityLog, 'utf8');
        const lines = log.split(/\r?\n/);
        const highlights = lines.filter((line) =>
          /exception|error|sampled-skeleton|animator has no|executeMethod|aborting batchmode/i.test(
            line,
          ),
        );
        tail = [...highlights.slice(-40), '--- final Unity lines ---', ...lines.slice(-30)].join(
          '\n',
        );
      } catch {
        // The process error is still useful if Unity failed before opening a log.
      }
      throw new Error(
        `${spec.unit}: Unity sampled-skeleton export failed` +
          (tail ? `\n--- Unity log tail ---\n${tail}` : ''),
        { cause: error },
      );
    }
    const samples = JSON.parse(await readFile(samplePath, 'utf8'));
    validateSamples(samples, timings, spec.unit);
    const result = await samplesToGlb(samples, spec, timings);
    console.log(
      `${spec.unit}: ${result.byteLength.toLocaleString()}-byte Unity-sampled skeletal GLB`,
    );
    return result;
  } finally {
    // Unity can release its redirected log a fraction after the process exits
    // on Windows. Retrying keeps an otherwise-successful import from being
    // reported as a cleanup failure while still guaranteeing temp isolation.
    await rm(tempRoot, {
      recursive: true,
      force: true,
      maxRetries: 8,
      retryDelay: 125,
    });
  }
}

async function samplesToGlb(data, spec, timings) {
  const reflection = new THREE.Matrix4().makeScale(-1, 1, 1);
  const meshData = data.mesh;
  const boneNodeSet = new Set(meshData.boneNodes);
  const objects = data.nodes.map((record, index) => {
    const object = boneNodeSet.has(index) ? new THREE.Bone() : new THREE.Group();
    const safeName = THREE.PropertyBinding.sanitizeNodeName(record.name) || 'node';
    object.name = `n${index}_${safeName}`;
    return object;
  });
  const pathToIndex = new Map(data.nodes.map((record, index) => [record.path, index]));
  const parentIndices = new Array(objects.length).fill(-1);
  for (let index = 0; index < objects.length; index++) {
    const path = data.nodes[index].path;
    const slash = path.lastIndexOf('/');
    if (index > 0) {
      const parentPath = slash < 0 ? '' : path.slice(0, slash);
      const parent = pathToIndex.get(parentPath);
      if (parent === undefined) {
        throw new Error(`${spec.unit}: Unity node ${path} has no parent`);
      }
      parentIndices[index] = parent;
      objects[parent].add(objects[index]);
    }
    applyUnityLocal(data.restTransforms, index * 10, objects[index]);
  }

  const vertexCount = meshData.vertices.length / 3;
  const geometry = new THREE.BufferGeometry();
  const positions = new Float32Array(meshData.vertices.length);
  const normals = new Float32Array(meshData.normals.length);
  for (let index = 0; index < meshData.vertices.length; index += 3) {
    positions[index] = -meshData.vertices[index] * 100;
    positions[index + 1] = meshData.vertices[index + 1] * 100;
    positions[index + 2] = meshData.vertices[index + 2] * 100;
    normals[index] = -meshData.normals[index];
    normals[index + 1] = meshData.normals[index + 1];
    normals[index + 2] = meshData.normals[index + 2];
  }
  if (vertexCount > 65_535) {
    throw new Error(`${spec.unit}: sampled mesh exceeds 16-bit topology`);
  }
  const indices = new Uint16Array(meshData.triangles.length);
  for (let index = 0; index < indices.length; index += 3) {
    indices[index] = meshData.triangles[index];
    indices[index + 1] = meshData.triangles[index + 2];
    indices[index + 2] = meshData.triangles[index + 1];
  }
  geometry.setAttribute('position', new THREE.BufferAttribute(positions, 3));
  geometry.setAttribute('normal', new THREE.BufferAttribute(normals, 3));
  geometry.setAttribute('uv', new THREE.BufferAttribute(new Float32Array(meshData.uv), 2));
  geometry.setAttribute(
    'skinIndex',
    new THREE.BufferAttribute(new Uint16Array(meshData.boneIndices), 4),
  );
  geometry.setAttribute(
    'skinWeight',
    new THREE.BufferAttribute(new Float32Array(meshData.boneWeights), 4),
  );
  geometry.setIndex(new THREE.BufferAttribute(indices, 1));
  geometry.computeBoundingBox();
  geometry.computeBoundingSphere();

  const mesh = new THREE.SkinnedMesh(
    geometry,
    new THREE.MeshStandardMaterial({
      color: 0xffffff,
      roughness: 0.8,
      metalness: 0,
    }),
  );
  mesh.name = `${spec.unit}Mesh`;
  mesh.userData = {
    animationFormat: 'athena2-unity-sampled-skeleton-v1',
    sourceVertexCount: vertexCount,
    clipFrames: Object.fromEntries(CLIP_NAMES.map((name) => [name, timings[name].frames])),
  };
  const exportRoot = new THREE.Group();
  exportRoot.name = `${spec.unit}UnitySampled`;
  exportRoot.add(objects[0]);
  exportRoot.add(mesh);
  // Match the model-node correction used by the normal FBX importer. Keeping
  // this above the resolved Unity hierarchy preserves every sampled local
  // transform while fixing source rigs whose FBX omits its authored up/facing
  // parent transform.
  if (spec.rotateX !== 0) {
    exportRoot.quaternion.premultiply(
      new THREE.Quaternion().setFromAxisAngle(new THREE.Vector3(1, 0, 0), spec.rotateX),
    );
  }
  if (spec.rotateY !== 0) {
    exportRoot.quaternion.premultiply(
      new THREE.Quaternion().setFromAxisAngle(new THREE.Vector3(0, 1, 0), spec.rotateY),
    );
  }
  exportRoot.updateMatrixWorld(true);

  const boneInverses = meshData.boneNodes.map((_, bone) => {
    const source = new THREE.Matrix4().fromArray(meshData.bindPoses, bone * 16);
    const converted = reflection.clone().multiply(source).multiply(reflection);
    converted.elements[12] *= 100;
    converted.elements[13] *= 100;
    converted.elements[14] *= 100;
    return converted;
  });
  const skeleton = new THREE.Skeleton(
    meshData.boneNodes.map((index) => objects[index]),
    boneInverses,
  );
  mesh.bind(skeleton, new THREE.Matrix4());
  mesh.normalizeSkinWeights();

  const animatedNodes = new Set(meshData.boneNodes);
  for (const boneNode of meshData.boneNodes) {
    let parent = parentIndices[boneNode];
    while (parent >= 0) {
      animatedNodes.add(parent);
      parent = parentIndices[parent];
    }
  }
  const clips = data.clips.map((sampleClip) => sampledClip(sampleClip, objects, animatedNodes));
  exportRoot.animations = clips;
  const glb = await new GLTFExporter().parseAsync(exportRoot, {
    animations: clips,
    binary: true,
    onlyVisible: false,
  });
  if (!(glb instanceof ArrayBuffer)) {
    throw new Error(`${spec.unit}: Unity sampled-skeleton exporter returned no GLB`);
  }
  return glb;
}

function sampledClip(sampleClip, objects, animatedNodes) {
  const sampleCount = sampleClip.authoredFrames + 1;
  const times = new Float32Array(sampleCount);
  for (let frame = 0; frame < sampleCount; frame++) {
    times[frame] = frame / sampleClip.frameRate;
  }
  const tracks = [];
  for (const node of animatedNodes) {
    const positions = new Float32Array(sampleCount * 3);
    const quaternions = new Float32Array(sampleCount * 4);
    const scales = new Float32Array(sampleCount * 3);
    for (let frame = 0; frame < sampleCount; frame++) {
      const offset = (frame * objects.length + node) * 10;
      positions[frame * 3] = -sampleClip.transforms[offset] * 100;
      positions[frame * 3 + 1] = sampleClip.transforms[offset + 1] * 100;
      positions[frame * 3 + 2] = sampleClip.transforms[offset + 2] * 100;
      // Reflection across X converts Unity's handedness: R * rotation * R.
      quaternions[frame * 4] = sampleClip.transforms[offset + 3];
      quaternions[frame * 4 + 1] = -sampleClip.transforms[offset + 4];
      quaternions[frame * 4 + 2] = -sampleClip.transforms[offset + 5];
      quaternions[frame * 4 + 3] = sampleClip.transforms[offset + 6];
      scales[frame * 3] = sampleClip.transforms[offset + 7];
      scales[frame * 3 + 1] = sampleClip.transforms[offset + 8];
      scales[frame * 3 + 2] = sampleClip.transforms[offset + 9];
    }
    addCompactVectorTrack(tracks, objects[node], 'position', times, positions, POSITION_EPSILON);
    addCompactQuaternionTrack(tracks, objects[node], times, quaternions);
    addCompactVectorTrack(tracks, objects[node], 'scale', times, scales, SCALE_EPSILON);
  }
  return new THREE.AnimationClip(
    sampleClip.name,
    sampleClip.authoredFrames / sampleClip.frameRate,
    tracks,
  );
}

function addCompactVectorTrack(tracks, object, property, times, values, epsilon) {
  const rest = object[property];
  let maxRest = 0;
  let maxFirst = 0;
  for (let frame = 0; frame < times.length; frame++) {
    for (let component = 0; component < 3; component++) {
      maxRest = Math.max(
        maxRest,
        Math.abs(values[frame * 3 + component] - rest.getComponent(component)),
      );
      maxFirst = Math.max(maxFirst, Math.abs(values[frame * 3 + component] - values[component]));
    }
  }
  if (maxRest <= epsilon) return;
  const trackTimes =
    maxFirst <= epsilon ? new Float32Array([times[0], times[times.length - 1]]) : times;
  const trackValues =
    maxFirst <= epsilon ? new Float32Array([...values.slice(0, 3), ...values.slice(0, 3)]) : values;
  tracks.push(new THREE.VectorKeyframeTrack(`${object.name}.${property}`, trackTimes, trackValues));
}

function addCompactQuaternionTrack(tracks, object, times, values) {
  const rest = object.quaternion;
  const first = new THREE.Quaternion().fromArray(values, 0);
  const current = new THREE.Quaternion();
  const previous = new THREE.Quaternion().copy(first);
  let maxRest = 0;
  let maxFirst = 0;
  for (let frame = 0; frame < times.length; frame++) {
    current.fromArray(values, frame * 4);
    if (frame > 0 && current.dot(previous) < 0) {
      for (let component = 0; component < 4; component++) {
        values[frame * 4 + component] *= -1;
      }
      current.fromArray(values, frame * 4);
    }
    maxRest = Math.max(maxRest, 1 - Math.abs(current.dot(rest)));
    maxFirst = Math.max(maxFirst, 1 - Math.abs(current.dot(first)));
    previous.copy(current);
  }
  if (maxRest <= QUATERNION_EPSILON) return;
  const trackTimes =
    maxFirst <= QUATERNION_EPSILON ? new Float32Array([times[0], times[times.length - 1]]) : times;
  const trackValues =
    maxFirst <= QUATERNION_EPSILON
      ? new Float32Array([...values.slice(0, 4), ...values.slice(0, 4)])
      : values;
  tracks.push(
    new THREE.QuaternionKeyframeTrack(`${object.name}.quaternion`, trackTimes, trackValues),
  );
}

function applyUnityLocal(values, offset, object) {
  object.position.set(-values[offset] * 100, values[offset + 1] * 100, values[offset + 2] * 100);
  object.quaternion.set(
    values[offset + 3],
    -values[offset + 4],
    -values[offset + 5],
    values[offset + 6],
  );
  object.scale.set(values[offset + 7], values[offset + 8], values[offset + 9]);
}

function validateTimings(timings, unit) {
  for (const name of CLIP_NAMES) {
    const timing = timings?.[name];
    if (!(timing?.frames > 1) || !(timing.frameRate > 0)) {
      throw new Error(`${unit}: invalid authored ${name} timing`);
    }
  }
}

function validateSamples(samples, timings, unit) {
  if (samples.unityVersion !== UNITY_VERSION) {
    throw new Error(
      `${unit}: sampled with Unity ${samples.unityVersion}, expected ${UNITY_VERSION}`,
    );
  }
  if (!Array.isArray(samples.nodes) || samples.nodes.length === 0) {
    throw new Error(`${unit}: Unity returned no sampled hierarchy`);
  }
  const mesh = samples.mesh;
  const vertexCount = (mesh?.vertices?.length ?? 0) / 3;
  if (
    !Number.isInteger(vertexCount) ||
    vertexCount <= 0 ||
    mesh.normals.length !== vertexCount * 3 ||
    mesh.uv.length !== vertexCount * 2 ||
    mesh.boneIndices.length !== vertexCount * 4 ||
    mesh.boneWeights.length !== vertexCount * 4 ||
    mesh.bindPoses.length !== mesh.boneNodes.length * 16
  ) {
    throw new Error(`${unit}: Unity returned malformed sampled mesh data`);
  }
  for (const name of CLIP_NAMES) {
    const clip = samples.clips?.find((candidate) => candidate.name === name);
    const timing = timings[name];
    if (
      clip?.authoredFrames !== timing.frames ||
      clip?.frameRate !== timing.frameRate ||
      clip?.transforms?.length !== (timing.frames + 1) * samples.nodes.length * 10
    ) {
      throw new Error(`${unit}: Unity returned malformed ${name} samples`);
    }
  }
}

function unityAssetPath(name) {
  return `Assets/Source/${basename(name)}`.replaceAll('\\', '/');
}

async function firstAvailablePath(candidates) {
  for (const candidate of candidates) {
    if (!candidate) continue;
    try {
      await access(candidate);
      return candidate;
    } catch (error) {
      if (error.code !== 'ENOENT') throw error;
    }
  }
  return null;
}

async function runProcess(command, args) {
  await new Promise((resolvePromise, reject) => {
    const child = spawn(command, args, {
      stdio: ['ignore', 'inherit', 'inherit'],
      windowsHide: true,
    });
    child.on('error', reject);
    child.on('exit', (code, signal) => {
      if (code === 0) resolvePromise();
      else {
        reject(new Error(`${command} exited with ${code ?? `signal ${signal ?? 'unknown'}`}`));
      }
    });
  });
}

function installBrowserShims() {
  globalThis.window ??= globalThis;
  globalThis.document ??= {
    createElementNS() {
      const listeners = new Map();
      const image = {
        addEventListener(type, listener) {
          listeners.set(type, listener);
        },
        removeEventListener(type) {
          listeners.delete(type);
        },
      };
      Object.defineProperty(image, 'src', {
        set() {
          queueMicrotask(() => listeners.get('load')?.());
        },
      });
      return image;
    },
  };
  globalThis.FileReader ??= class FileReader {
    result = null;
    onloadend = null;

    readAsArrayBuffer(blob) {
      void blob.arrayBuffer().then((result) => {
        this.result = result;
        this.onloadend?.({ target: this });
      });
    }

    readAsDataURL(blob) {
      void blob.arrayBuffer().then((result) => {
        this.result = `data:${blob.type || 'application/octet-stream'};base64,${Buffer.from(result).toString('base64')}`;
        this.onloadend?.({ target: this });
      });
    }
  };
}
